### PR TITLE
Add flags.deprecate_package_materialization_builtin_override

### PR DESCRIPTION
**Pull request by MichelleArk**
resolves https://github.com/dbt-labs/dbt-core/issues/10007

### Problem

Materializations defined in third-party packages are included in the search order for materialization macros. This is difficult to detect, difficult to prevent, and it works differently from other macros.

We should remove third-party packages from the search order.

This change needs to go behind a behavior-change flag:[Legacy behaviors | dbt Developer Hub](https://docs.getdbt.com/reference/global-configs/legacy-behaviors) . See https://github.com/dbt-labs/dbt-core/pull/9366 for an implementation of a behavior-change flag.

### Solution

Introduce flags.deprecate_package_materialization_builtin_override.

when the flag is False (legacy behaviour):
* fire a deprecation event and warning and allow the override
when the flag is True (soon-to-be default behaviour)
* get the next candidate, only searching for candidates with core and root localities

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

### CODE REVIEW

File: core/dbt/cli/flags.py [hunk 1]
```diff
@@ -355,7 +355,7 @@ def set_common_global_flags(self):
     # This is here to prevent mypy from complaining about all of the
     # attributes which we added dynamically.
     def __getattr__(self, name: str) -> Any:
-        return super().__get_attribute__(name)  # type: ignore
+        return super().__getattribute__(name)  # type: ignore
```
Comment (MichelleArk) [inline core/dbt/cli/flags.py:358]: This appears to have been broken all along, but was never reached in testing

File: core/dbt/contracts/graph/manifest.py [hunk 2]
```diff
@@ -946,11 +960,17 @@ def find_materialization_macro_by_name(
...
+            # preserve legacy behaviour - allow materialization override
+            if get_flags().deprecate_package_materialization_builtin_override is False:
+                deprecations.warn(
+                    "package-materialization-override",
+                    package_name=materialization_candidate.macro.package_name,
+                    materialization_name=materialization_name,
+                )
+            else:
+                materialization_candidate = candidates.last_candidate(
+                    valid_localities=[Locality.Core, Locality.Root]
```
Comment (MichelleArk) [inline core/dbt/contracts/graph/manifest.py:972]: was not able to do this lower-level using the `_find_macros_by_name` `filter` parameter because that filter only has a view of the macro candidate being iterated over. For this check, we need to exclude imported materialization macros only if the materialization is builtin in order to continue supporting custom materializations from packages (tested)

**Comment by MichelleArk** [comment]
Note: we'd like to backport these changes but it will very likely need to be manual since ProjectFlags does not exist in 1.6 or 1.7

File: .changes/unreleased/Features-20240422-173703.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+kind: Features
+body: Add flags.deprecate_package_materialization_builtin_override
```
Comment (emmyoop) [inline .changes/unreleased/Features-20240422-173703.yaml:2]: Can you reword this to be more human readable / not code?

**Review by emmyoop — APPROVED** [review]
Looks good - just a changelog comment.

**Comment by MichelleArk** [comment]
codecov/project is hanging waiting for a response but I can see the results passing on: https://app.codecov.io/gh/dbt-labs/dbt-core/pull/9956
Screenshot 2024-04-23 at 3 54 03 PM
